avro_schema/utils.lua

```diff
@@ -53,8 +53,137 @@ function has_only(t, key)
+    assert(table_contains(keys, 'clear') == false)
+    local push_str = [[
+    return function(stack, VARS)
+        stack.len = stack.len + 1
+        SET_VALS
+    end]]
+    push_str = push_str:gsub('VARS',
+        table.concat(
+            map(function (i, _) return 'v' .. i end, keys),
+            ',')
+        )
+    push_str = push_str:gsub('SET_VALS',
+        table.concat(
+            map(function (i, key)
+                return ('stack.%s[stack.len] = v%d'):format(key, i)
+            end, keys),
+            '\n')
+        )
+    local pop_str = [[
+    return function (stack)
+        local res = stack:get(stack.len)
+        stack:remove_last()
+        return res
+    end
+    ]]
+    local remove_last_str = [[
+    return function (stack)
+        stack.len = stack.len - 1
+        assert(stack.len >= 0)
+    end
+    ]]
+    local get_str = [=[
+    return function (stack, pos)
+        if pos <= 0 or pos > stack.len then
+            error(("Attempt to get element %d from stack of size %d %s"):format(
+                pos, stack.len, require('yaml').encode(debug.traceback())))
+        end
+        local ret = {
+            RET_TAB_VALS
+        }
+        return ret
+    end]=]
+    get_str = get_str:gsub('RET_TAB_VALS', 
+        table.concat(
+            map(function (i, key)
+                return ('%s = stack.%s[pos]'):format(key, key)
+            end, keys),
+            ',\n')
+        )
+    local clear_str = [[
+    return function (stack)
+        stack.len = 0
+        CLEAR_TABLES
+    end
+    ]]
+    clear_str = clear_str:gsub('CLEAR_TABLES', 
+        table.concat(
+            map(function (_, key)
+                return ('table.clear(stack.%s)'):format(key)
+            end, keys),
+            '\n')
+        )
+    local stack = {
+        len = 0,
+        push = loadstring(push_str, 'avro.fstack.push')(),
+        pop = loadstring(pop_str, 'avro.fstack.pop')(),
+        remove_last = loadstring(remove_last_str, 'avro.fstack.remove_last')(),
+        get = loadstring(get_str, 'avro.fstack.get')(),
+        clear = loadstring(clear_str, 'avro.fstack.clear')(),
+    }
+    -- Gen setters/getters.
+    -- Setters/getters are a bit slower than raw access.
+    for _, key in pairs(keys) do
+        local setter_str = ([[
+            return function (stack, val)
+                assert(stack.len > 0)
+                stack.VAL_NAME[stack.len] = val
+            end]]):gsub('VAL_NAME', key)
+        stack['set_' .. key] = loadstring(setter_str,
+            'avro.fstack.set_' .. key)()
+        local getter_str = ([[
+            return function (stack, val)
+                assert(stack.len > 0)
+                return stack.VAL_NAME[stack.len]
+            end]]):gsub('VAL_NAME', key)
+        stack['get_' .. key] = loadstring(getter_str,
+            'avro.fstack.get_' .. key)()
+    end
+    -- Gen blank arrays for all stack keys.
+    for _, key in pairs(keys) do
```

Comment:
I prefer use `ipairs` when a table is known to be a list for readability.

Reply:
fixed